cirq-core/cirq/transformers/merge_single_qubit_gates.py

```diff
@@ -0,0 +1,132 @@
+
+from cirq import protocols, circuits
+from cirq.transformers.analytical_decompositions import single_qubit_decompositions
+from cirq.transformers import transformer_api, transformer_primitives, merge_k_qubit_gates
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@transformer_api.transformer
+def merge_single_qubit_gates_to_phased_x_and_z(
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    atol: float = 1e-8,
+) -> 'cirq.Circuit':
+    """Replaces runs of single qubit rotations with `cirq.PhasedXPowGate` and `cirq.ZPowGate`.
+
+    Specifically, any run of non-parameterized single-qubit unitaries will be replaced by an
+    optional PhasedX operation followed by an optional Z operation.
+
+    Args:
+        circuit: Input circuit to transform. It will not be modified.
+        context: `cirq.TransformerContext` storing common configurable options for transformers.
+        atol: Absolute tolerance to angle error. Larger values allow more negligible gates to be
+            dropped, smaller values increase accuracy.
+
+    Returns:
+        Copy of the transformed input circuit.
+    """
+
+    def rewriter(op: 'cirq.CircuitOperation') -> 'cirq.OP_TREE':
+        return [
+            g(op.qubits[0])
+            for g in single_qubit_decompositions.single_qubit_matrix_to_phased_x_z(
+                protocols.unitary(op), atol
+            )
+        ]
+
+    return merge_k_qubit_gates.merge_k_qubit_unitaries(
+        circuit, k=1, context=context, rewriter=rewriter
+    )
+
+
+@transformer_api.transformer
+def merge_single_qubit_gates_to_phxz(
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    atol: float = 1e-8,
+) -> 'cirq.Circuit':
+    """Replaces runs of single qubit rotations with a single optional `cirq.PhasedXZGate`.
+
+    Specifically, any run of non-parameterized single-qubit unitaries will be replaced by an
+    optional PhasedXZ.
+
+    Args:
+        circuit: Input circuit to transform. It will not be modified.
+        context: `cirq.TransformerContext` storing common configurable options for transformers.
+        atol: Absolute tolerance to angle error. Larger values allow more negligible gates to be
+            dropped, smaller values increase accuracy.
+
+    Returns:
+        Copy of the transformed input circuit.
+    """
+
+    def rewriter(op: 'cirq.CircuitOperation') -> 'cirq.OP_TREE':
+        gate = single_qubit_decompositions.single_qubit_matrix_to_phxz(protocols.unitary(op), atol)
+        return gate(op.qubits[0]) if gate else []
+
+    return merge_k_qubit_gates.merge_k_qubit_unitaries(
+        circuit, k=1, context=context, rewriter=rewriter
+    )
+
+
+@transformer_api.transformer
+def merge_single_qubit_moments_to_phxz(
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    atol: float = 1e-8,
+) -> 'cirq.Circuit':
```

Comment:
Why do we need a moment based transformer in addition to a gate based one ?

Reply:
See the tests for an example use case. 

The idea is that if I give you a moment preserving circuit with alternating layers and 1 and 2q gates that you need to convert to a target gateset, during conversion you will often end up in a situation where there are two adjacent moments of single qubit gates. Calling an operation based transformer to merge the single qubit gates is not sufficient as it can result in 2 separate moments in the final result, whereas a moment based transformer will guarantee that the two moments are merged into one.